diff --git a/scp/1_samtools.cpp b/scp/1_samtools.cpp
index 3bc7057..6801a3b 100644
--- a/scp/1_samtools.cpp
+++ b/scp/1_samtools.cpp
@@ -23,7 +23,7 @@ int samtools(string working_dir, string input_bam, string chr, string start, str
     //std::cin.tie(0);
     
     string sys;
-    sys="samtools view -q "+MAPQ+" -F 0x100 -F 0x200  -F 0x400 -T "+fasta+" "+input_bam+" "+chr+":"+start+"-"+end+" |sed -e 's/[ ][ ]*/_/g'  > "+working_dir+"region.sam";
+    sys="samtools view -q "+MAPQ+" -F 0x100 -F 0x200 -F 0x400 -F 0x800 -T "+fasta+" "+input_bam+" "+chr+":"+start+"-"+end+" |sed -e 's/[ ][ ]*/_/g'  > "+working_dir+"region.sam";
     
     char *syst = new char[sys.length()+1];
     strcpy(syst, sys.c_str());
diff --git a/scp/7_FP_ex.cpp b/scp/7_FP_ex.cpp
index ed3f46b..8af6b6b 100755
--- a/scp/7_FP_ex.cpp
+++ b/scp/7_FP_ex.cpp
@@ -642,6 +642,9 @@ int fp_ex(string WD_dir, string fasta, string chr, string t, int tsd_index){
             //file20.close();
             file21.close();
             file21.clear();
+            
+            // Delete the intermediate file to avoid exhausting inodes.
+            remove(syst_ref_junc_file);
         }
         
         for(int w=0;w!=line_tsd;++w){
diff --git a/scp/8_calling.cpp b/scp/8_calling.cpp
index 6c3fd77..3b65ebd 100755
--- a/scp/8_calling.cpp
+++ b/scp/8_calling.cpp
@@ -1467,7 +1467,7 @@ int calling(string WD_dir, string t, int tsd_index){
                                 file4<<seq5[n];
                             }
 //5bp
-                            for(int n=BIN_buff;n!=info_line[read_seq][2].length()-1-BIN_buff;++n){
+                            for(int n=BIN_buff;n<info_line[read_seq][2].length()-1-BIN_buff;++n){
                                 file4<<seq_line[n];
                             }
                             
@@ -1482,7 +1482,7 @@ int calling(string WD_dir, string t, int tsd_index){
                                 file4<<seq3[n];
                             }
 //5bp
-                            for(int n=1+BIN_buff;n!=info_line[read_seq][2].length()-1-BIN_buff;++n){
+                            for(int n=1+BIN_buff;n<info_line[read_seq][2].length()-1-BIN_buff;++n){
                                 file4<<seq_line[n];
                             }
                             
@@ -1920,7 +1920,7 @@ int calling(string WD_dir, string t, int tsd_index){
                                 file4<<seq5[n];
                             }
                             
-                            for(int n=BIN_buff;n!=info_line[read_seq][2].length()-1-BIN_buff;++n){
+                            for(int n=BIN_buff;n<info_line[read_seq][2].length()-1-BIN_buff;++n){
                                 file4<<seq_line[n];
                             }
                             
@@ -1935,7 +1935,7 @@ int calling(string WD_dir, string t, int tsd_index){
                                 file4<<seq3[n];
                             }
                             
-                            for(int n=1+BIN_buff;n!=info_line[read_seq][2].length()-1-BIN_buff;++n){
+                            for(int n=1+BIN_buff;n<info_line[read_seq][2].length()-1-BIN_buff;++n){
                                 file4<<seq_line[n];
                             }
                             
