diff --git a/scp/1_samtools.cpp b/scp/1_samtools.cpp
index 3bc7057..6801a3b 100644
--- a/scp/1_samtools.cpp
+++ b/scp/1_samtools.cpp
@@ -23,7 +23,7 @@ int samtools(string working_dir, string input_bam, string chr, string start, str
     //std::cin.tie(0);
     
     string sys;
-    sys="samtools view -q 10 -F 0x100 -F 0x200  -F 0x400 -T "+fasta+" "+input_bam+" "+chr+":"+start+"-"+end+" |sed -e 's/[ ][ ]*/_/g'  > "+working_dir+"region.sam";
+    sys="samtools view -q 10 -F 0x100 -F 0x200 -F 0x400 -F 0x800 -T "+fasta+" "+input_bam+" "+chr+":"+start+"-"+end+" |sed -e 's/[ ][ ]*/_/g'  > "+working_dir+"region.sam";
     
     char *syst = new char[sys.length()+1];
     strcpy(syst, sys.c_str());
